File: AppTemplate/native/ios/Sources/Plugins/Notification/NotificationPlugin.swift
import Foundation
import UIKit
import UserNotifications

public class NotificationPlugin: SweetJuicePlugin {
    private var container: UIViewController?

    public init() {}

    public func getDomain() -> String {
        return "notification"
    }

    public func onAttach(container: UIViewController) {
        self.container = container
    }

    public func handleAction(action: String, jsonArgs: String) -> String {
        guard let data = jsonArgs.data(using: .utf8),
              let args = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return "{\"error\":\"Invalid JSON payload\"}"
        }

        if action == "post" {
            let id = args["id"] as? Int ?? Int(Date().timeIntervalSince1970)
            let title = args["title"] as? String ?? ""
            let body = args["body"] as? String ?? ""

            postNotification(id: id, title: title, body: body)
            return "{\"status\":\"posted\", \"id\":\(id)}"
        }

        if action == "cancel" {
            let id = args["id"] as? Int ?? -1
            UNUserNotificationCenter.current().removePendingNotificationRequests(withIdentifiers: ["\(id)"])
            return "{\"status\":\"cancelled\"}"
        }

        return "{\"error\":\"Unknown action\"}"
    }

    private func postNotification(id: Int, title: String, body: String) {
        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default

        let trigger = UNTimeIntervalNotificationTrigger(timeInterval: 1, repeats: false)
        let request = UNNotificationRequest(identifier: "\(id)", content: content, trigger: trigger)

        UNUserNotificationCenter.current().add(request) { error in
            if let error = error {
                print("Error posting notification: \(error.localizedDescription)")
            }
        }
    }
}

File: AppTemplate/.plugins/android/com/pkg/notifications/NotificationPlugin.java
package com.juicepackage.notifications;

import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import androidx.core.app.NotificationCompat;

import com.juiceplugin.SweetJuicePlugin;

import org.json.JSONException;
import org.json.JSONObject;

/**
 * NotificationPlugin allows Go to post system notifications.
 */
public class NotificationPlugin implements SweetJuicePlugin {
    private Context mContext;

    @Override
    public String getDomain() { return "notification"; }

    @Override
    public void onAttach(Context context) { 
        this.mContext = context; 
    }

    @Override
    public String handleAction(String action, String jsonArgsPayload) {
        try {
            JSONObject args = new JSONObject(jsonArgsPayload);

            if ("post".equals(action)) {
                int id = args.optInt("id", (int) System.currentTimeMillis());
                String title = args.optString("title", "");
                String body = args.optString("body", "");
                String channelId = args.optString("channel_id", "default");
                String channelName = args.optString("channel_name", "Default");
                String importance = args.optString("importance", "DEFAULT");

                postNotification(id, title, body, channelId, channelName, importance);
                return "{\"status\":\"posted\", \"id\":" + id + "}";
            }

            if ("cancel".equals(action)) {
                int id = args.optInt("id", -1);
                NotificationManager manager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
                manager.cancel(id);
                return "{\"status\":\"cancelled\"}";
            }

        } catch (JSONException e) {
            return "{\"error\":\"Invalid JSON payload\"}";
        }

        return "{\"error\":\"Unknown action\"}";
    }

    private void postNotification(int id, String title, String body, String channelId, String channelName, String importanceStr) {
        NotificationManager manager = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            int importance = NotificationManager.IMPORTANCE_DEFAULT;
            switch (importanceStr) {
                case "HIGH": importance = NotificationManager.IMPORTANCE_HIGH; break;
                case "LOW": importance = NotificationManager.IMPORTANCE_LOW; break;
                case "MIN": importance = NotificationManager.IMPORTANCE_MIN; break;
            }
            NotificationChannel channel = new NotificationChannel(channelId, channelName, importance);
            manager.createNotificationChannel(channel);
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(mContext, channelId)
                .setSmallIcon(mContext.getApplicationInfo().icon)
                .setContentTitle(title)
                .setContentText(body)
                .setPriority(NotificationCompat.PRIORITY_DEFAULT)
                .setAutoCancel(true);

        manager.notify(id, builder.build());
    }

    @Override public void onRequestPermissionsResult(int rc, String[] p, int[] g) {}
    @Override public void onActivityResult(int r, int rc, Intent d) {}
}
